### DISCUSSION TITLE
非网页程序员背景问几个小问题

### DISCUSSION THREAD

**Discussion by kenlsoft** [HVML/PurC#38] [General]
1. 示例文档中，一开头的"!DOCTYPE hvml"是因为要兼容HTML或者W3C的标准吗？因为有一个最简的hello world程序中，依然有这个标签。
2. 在解析过程中，是不是所有HVML独特的标签都按照HVML解释器处理，所有的外部标签，比如HTML定义的，都原模原样的复制到了输出格式中（当然含有变量或者运算公式的会被解释）？是否会根据target=""来选择相应的输出处理程序，只不过目前仅支持html和void?
3. 输出文档的目的可以理解为方便网页渲染；运行一次程序就能得到一个可供网页浏览器打开的页面。那么在那种交互式的网页中，会怎么处理？因为程序还没退出（等事件中），所以这个最终的文档还没产生，是不是会产生一个中间文档？
4. 计算器的例子中，用以处理排版等样式的语句依然与代码逻辑在一起(div等），这部分有必要分开吗？

**Comment by VincentWei** [comment]
> 示例文档中，一开头的"!DOCTYPE hvml"是因为要兼容HTML或者W3C的标准吗？因为有一个最简的hello world程序中，依然有这个标签。

这个标签是可选的的。目前有两个作用：

1. 用于 file 等命令或者函数库可以快速判断文件类型。
2. 用于定义预加载的共享对象及其名称。

> 在解析过程中，是不是所有HVML独特的标签都按照HVML解释器处理，所有的外部标签，比如HTML定义的，都原模原样的复制到了输出格式中（当然含有变量或者运算公式的会被解释）？是否会根据target=""来选择相应的输出处理程序，只不过目前仅支持html和void?

您的理解非常正确，外部元素中包含的表达式（属性值或者内容）都会被求值，然后再将其克隆到目标文档。目前只支持 HTML 和 VOID 两种类型，未来会根据需要扩展。

> 输出文档的目的可以理解为方便网页渲染；运行一次程序就能得到一个可供网页浏览器打开的页面。那么在那种交互式的网页中，会怎么处理？因为程序还没退出（等事件中），所以这个最终的文档还没产生，是不是会产生一个中间文档？

HVML 程序的执行分为几个阶段，其中有一个阶段称为首次运行（first-run），在首次运行结束后，会将文档发送给渲染器做一次渲染，如果注册有观察者，则会进入观察（observing）阶段，否则会进入清理（cleanup）阶段。在观察阶段，HVML 解释器和渲染器之间进行常规的事件交互和文档更新操作。

> 计算器的例子中，用以处理排版等样式的语句依然与代码逻辑在一起(div等），这部分有必要分开吗？

如果需要将代码和文档片段分离，可以使用 `archetype` 从外部文件（或 URL）中获取模板片段。HVML 甚至支持使用`define/include` 等元素从外部文件（或 URL）中获取程序片段。
